evals/solvers/providers/fixie/local_gpu_solver.py

```diff
@@ -110,18 +133,121 @@ def solver_initializer(
     if torch.cuda.is_available():
         device = torch.device("cuda", rank)
     else:
-        raise ValueError("Only GPU mode is supported by LocalGPUSolver, but no GPUs were found")
+        device = torch.device("cpu")
 
-    global pipe
+    global pipe, collator
 
     pipe = transformers.pipeline(
-        model=model, trust_remote_code=True, device=device, **extra_options
+        model=model,
+        trust_remote_code=True,
+        device=device,
+        torch_dtype=torch.bfloat16,
+        **extra_options,
     )
+    pipe.tokenizer.padding_side = "left"
+
+    collator = DataCollatorForSeq2SeqWithAudio(tokenizer=pipe.tokenizer)
 
-    # Let the other initializers start now that the download has finished
-    for i in range(1, world_size):
-        rank_queue.put(i)
+    if rank == 0:
+        # Let the other initializers start now that the download has finished
+        for i in range(1, world_size):
+            rank_queue.put(i)
 
 
 def solver_worker(inputs: Dict[str, Any]):
-    return pipe(inputs)
+    prepped = [pipe.preprocess(item) for item in inputs]
+    prepped = [
+        {k: v.to(pipe.model.device).squeeze(0) for k, v in sample.items()} for sample in prepped
+    ]
+
+    batch = collator(prepped)
+    batch = {k: v.to(pipe.model.device) for k, v in batch.items() if v is not None}
```

Comment:
@liPatrick sorry I didn't end up using your code, but can you take a look to verify how the batching is happening here and if you had found better ideas?

Reply:
Yeah, this looks right. One question I do have, is how do you ensure the input lengths for each bath are similar?

Reply:
I don't. That seems very hard to optimize.

Actually what matters most IMO is not the input length, but the output lengths as that impacts generation time way more than input length variations. And output length seems impossible to determine ahead of time.